fenix/app/src/main/java/org/mozilla/fenix/ext/Activity.kt

```diff
@@ -174,3 +178,54 @@ const val REQUEST_CODE_BROWSER_ROLE = 1
 const val SETTINGS_SELECT_OPTION_KEY = ":settings:fragment_args_key"
 const val SETTINGS_SHOW_FRAGMENT_ARGS = ":settings:show_fragment_args"
 const val DEFAULT_BROWSER_APP_OPTION = "default_browser"
+const val EXTERNAL_APP_BROWSER_INTENT_SOURCE = "CUSTOM_TAB"
+
+/**
+ * Depending on the [Activity], maybe derive the source of the given [intent].
+ *
+ * @param intent the [SafeIntent] to derive the source from.
+ */
+fun Activity.getIntentSource(intent: SafeIntent): String? = when (this) {
+    is ExternalAppBrowserActivity -> EXTERNAL_APP_BROWSER_INTENT_SOURCE
+    else -> getHomeIntentSource(intent)
+}
```

Comment:
🚗 Instead of using `else` statements here, I'm wondering if we could use sealed interfaces to let these extracted functions (in all of your PRs) take advantage of compile-time type checking. For example:

```
sealed interface IntentSource

open class HomeActivity : IntentSource
open class ExternalAppBrowserActivity : IntentSource

fun IntentSource.getStringId(intent: SafeIntent): String? = when (this) {
    is ExternalAppBrowserActivity -> EXTERNAL_APP_BROWSER_INTENT_SOURCE
    is HomeActivity -> getHomeIntentSource(intent)
}
```

to be clear, I am not at all sure that this would work especially considering the subtypes are open. I think at the very least it would require moving some classes around to be in the same package.

At the very least, I think it might be beneficial to to handle each branch with the specific subtype and use the `else` case as an error case.

Reply:
Thanks for the suggestion @MatthewTighe, this is a great idea. As this also affects the other functions refactored as part of this project, I've created a new [bug](https://bugzilla.mozilla.org/show_bug.cgi?id=1873457) to track this work

Reply:
Update: I'm actually aiming to consolidate this into a single Activity (once these initial project PRs are merged) which will negate the need for this - stay tuned 🙂